### PR TITLE
Add a method to wait for PostgreSQL to show up

### PR DESCRIPTION
**Pull request by portante**
We add a method to wait for the configured `host`:`port` for the PostgreSQL database to accept socket connection.  We'll only wait at most N seconds before giving up (configurable).

We also tackle the following items:

 * Add doc strings and type hints where applicable
 * Consolidate the server configuration gathering to one location
   * Stream-lining the use of the `PbenchServerConfig` object
 * Keep the `site.ENABLE_USER_SITE` checks in `main()`
 * Narrow exception handling
 * Pass along the final gunicorn exit code

### CODE REVIEW

**Review by npalaska — APPROVED** [review]

File: lib/pbench/cli/server/shell.py
Comment (dbutenhof) [inline lib/pbench/cli/server/shell.py:156]: Intriguing. I do like the idea of essentially deferring the `crontab` startup until after we've started the server and initialized the DB; I'd been worrying a bit about that in the back of my mind.

Running `pbench-create-crontab` on each server startup seems excessive; should this be protected by a check for `f"{crontab_dir}/crontab"` existance?

Also, I've never tried to simply replace a running `crontab` by just installing a new one instead of deleting the old one first. Does that work reliably? (I don't see any explicit prohibition, but ...)

**Review by dbutenhof — DISMISSED** [review]

**Comment by dbutenhof** [comment]
> E           requests.exceptions.HTTPError: 503 Server Error: Service Unavailable for url: http://localhost:8080/api/v1/endpoints

So, your server is having a problem, even though the initial `curl` seemed to work. Ugh.

When I was debugging #3078 I added `Pipeline.gy` code to pull the `journalctl` output into the Jenkins log. We really need to work out a way to expose what's happening permanently. (I would probably have submitted the hack except that I couldn't find a reliable way to construct the container name due to an unwanted `" "` that kept creeping in.)

File: lib/pbench/cli/server/shell.py
Comment (webbnh) [inline lib/pbench/cli/server/shell.py:55]: This change seems like it creates more obfuscation than clarity or even convenience.

`add_to_path` isn't a sufficiently descriptive name, since it doesn't indicate whether the new path component is added to head or the tail of the path...whereas this detail was obvious looking at the old code.

Moreover, the operation isn't so complicated or specialized that we need to avoid duplicating it (and, the replacement seems to take up just as much space as the original at the call site).

If you really don't like using `+` for string concatenation, we could go with something like
```
        os.environ["PATH"] = ":".join((local, os.environ["PATH"]))
```
😉

File: lib/pbench/cli/server/shell.py
Comment (webbnh) [inline lib/pbench/cli/server/shell.py:76]: I don't think this function actually raises `ConnectionRefusedError` on failure:  other exceptions (from `create_connection()` might propagate out, but `ConnectionRefusedError` is caught and handled, and, if the timeout expires, then the function returns normally (with no return value).  (Did you mean to raise an exception after line 86?)

Also, not to quibble or anything, but `timeout` really refers to the number of attempts -- with a 1 second delay between them -- that the function will make...which isn't quite the same thing as the "number of seconds to wait".  If you want to implement what you've described, then line 80 should be something like
```
    end = time.time() + timeout
    while time.time() <= end:
```
🤷

File: lib/pbench/cli/server/shell.py
Comment (webbnh) [inline lib/pbench/cli/server/shell.py:91]: I'm not sure that containerization has anything to do with anything, here.  And, I'm not convinced that we want to claim to depend on running in/as a container.

The point to be made here is that the configuration file is in a known location.  It's too bad that we have to define it here, rather than pulling it in from a single point of truth (although, I guess it is possible that this _**is**_ the single point of truth for this...).

In any case, I would omit this whole sentence from the docstring, and perhaps add a description of what this function actually does (e.g., reads the server configuration, waits for the database service to become available, create the database, and so on)...or just a generic comment like, "Starts the Pbench Server".

**Review by webbnh — CHANGES_REQUESTED** [review]
I think that `wait_for_postgresql()` is failing to report failure.

Also, I don't think that the addition of `add_to_path()` is an improvement.

File: lib/pbench/cli/server/shell.py
Comment (portante) [reply lib/pbench/cli/server/shell.py:156]: Oh, wait, that should not be in this PR.  Sorry.  I'll fix.

**Review by portante — COMMENTED** [review]
Please review PR #3092 before continuing with this PR.

File: lib/pbench/cli/server/shell.py
Comment (webbnh) [inline lib/pbench/cli/server/shell.py:164]: Given the number of possibilities for failure, perhaps we should provide better feedback than this (yeah, the traceback probably has everything, but we don't really need/want a traceback here, especially if the message supplied the information in a nice format).  I think that those exception objects provide a nice formatting automagically:
```
    except (NoOptionError, NoSectionError) as exc:
        logger.error("Error fetching required configuration: {}", str(exc))
```

**Review by webbnh — CHANGES_REQUESTED** [review]
> Please review PR https://github.com/distributed-system-analysis/pbench/pull/3092 before continuing with this PR.

Done.  🙂 

Found a couple of bugs (I think...).  And, I don't think that it's a good idea for `shell.py` to (need to) be cognizant of what the database backend is.

And, I have a couple of suggestions.

**Review by webbnh — APPROVED** [review]
The fact that `wait_for_database()` requires the database URL to include a port number is a little...weird...but I guess it's OK...except for the fact that the functional tests seem to be failing.

**Review by riya-17 — DISMISSED** [review]

File: lib/pbench/cli/server/shell.py
Comment (dbutenhof) [inline lib/pbench/cli/server/shell.py:73]: This is "interesting" because the protocol doesn't formally require a port; the connector would normally default unless you're using an unusual port number. The problem of course is that we'd need to know the default port for each possible DB protocol in order to do the manual probe you're attempting here.

One solution would be a "formal" SQLAlchemy healthcheck, which basically amounts to something like doing the normal DB session init and trying a `select 1` query. (Which tests beyond just "something's responding on the port" to "yes, the DB is up and responding".) And it'd also work for `sqlite3` or any other lightweight local DB...

**Review by dbutenhof — DISMISSED** [review]
OK, but with a comment that would make this safer and more robust... at some cost. 😁

File: lib/pbench/cli/server/shell.py
Comment (portante) [reply lib/pbench/cli/server/shell.py:73]: I'd like to get this to land so we can move forward.

We are not talking about managing more than 2 configurations where the port would need to be specified: the first being the in-tree configuration used for the pbench-in-a-can container, the second being outside of the code base where it is maintained configured for the staging and production deployment (which may require a different port anyways).

This check is not trying to "know" about database semantics, just that the target database is reachable.  We could add a much deeper health check for the database, and that would probably be worth while, and could also handle the connectivity check.

But I don't see the value in waiting for that.

With the exception that I need to suss out what is happening with the functional tests.

Comment (dbutenhof) [reply lib/pbench/cli/server/shell.py:73]: Fair enough: agile is good, and a check is better than none. I had approved, and I'll resolve the conversation.

I'm also going to take a look at a general solution for pulling the container log journal in a minute, having finished a round of reviews for the morning. 😆 

In the meantime, adding `podman exec pbenchinacan_3089-pbenchserver journalctl` to the `Pipeline.gy` ought to do it. (I also changed the functional test launch to add `|| true` so that failure didn't abort, but you could also put the journalctl pull in the `post.always` section before the forced remove...)

**Review by dbutenhof — APPROVED** [review]
I assume this was just a rebase to get the log on failure; so while approval might be premature, I'll tick the box anyway.

**Comment by dbutenhof** [comment]
> Dec 02 15:52:18 pbenchinacan pbench-server[318]:   File "/opt/pbench-server/bin/pbench-server", line 10, in <module>
Dec 02 15:52:18 pbenchinacan pbench-server[318]:     sys.exit(main())
Dec 02 15:52:18 pbenchinacan pbench-server[318]:   File "/opt/pbench-server/lib/pbench/cli/server/shell.py", line 175, in main
Dec 02 15:52:18 pbenchinacan pbench-server[318]:     wait_for_database(db_uri, db_wait_timeout)
Dec 02 15:52:18 pbenchinacan pbench-server[318]:   File "/opt/pbench-server/lib/pbench/cli/server/shell.py", line 74, in wait_for_database
Dec 02 15:52:18 pbenchinacan pbench-server[318]:     end = time.time() + timeout
Dec 02 15:52:18 pbenchinacan pbench-server[318]: TypeError: unsupported operand type(s) for +: 'float' and 'str'

**Review by webbnh — DISMISSED** [review]
👍

**Review by portante — COMMENTED** [review]
The `shell` module feels like it needs a set of unit tests.  I can do that in this PR or in a PR immediately following this one.  Thoughts?

**Review by dbutenhof — APPROVED** [review]
> The shell module feels like it needs a set of unit tests. I can do that in this PR or in a PR immediately following this one. Thoughts?

That'd be lovely, but we don't need to add it for this PR.

**Review by npalaska — APPROVED** [review]
:+1:

**Review by webbnh — APPROVED** [review]
👍

**Comment by portante** [comment]
Thanks for merging, but I really wanted to handle the merge of this commit so that I could fix up the message.

**Comment by portante** [comment]
If we are going to allow relay on squash and merge then we need to also review the commit message.

**Comment by webbnh** [comment]
> Thanks for merging, but I really wanted to handle the merge of this commit so that I could fix up the message.

My apologies.  (Perhaps adding a note to that effect to the description might help prevent this in the future...or, we could decide on who has merge responsibilities.)

> If we are going to allow rely on squash and merge then we need to also review the commit message.

I think that squash-and-merge is helpful, in that it avoids the hassle and delay from the extra round of review (not to mention the temptation to make changes after the final approval).  Nevertheless, being in the habit of reviewing commit messages might be no bad idea!  Or, we could come up with some sort of process for doing squash-and-merge, like saying that the PR description should be used for the merge commit.

**Comment by portante** [comment]
I'd be satisfied with the PR description.

**Comment by dbutenhof** [comment]
FYI: I really want your deferred `crontab`. I've been trying to test some `PUT` fixes (and new functional tests), and everything's going insane because the crontab commands are now trying to initialize the DB and failing without exiting ... so nothing gets unpacked or indexed even once the server is up and running. It's bizarre; but an easy and sensible fix is to make sure they can't run until the server starts.

(Note: another solution might be to put the DB healthcheck inside the `Database.init_db()` method rather than in `shell.py` so it gets run by the tools, too... but I really like the idea of deferring crontab since they can't do anything anyway.)

**Comment by portante** [comment]
I'll post the deferred crontab solution next.

**Pull request merged**
